upstairs/src/dummy_downstairs_tests.rs

```diff
@@ -711,4 +722,999 @@ pub(crate) mod protocol_test {
 
         assert_eq!(ds1_message, ds1_message_second_time.unwrap());
     }
+
+    /// Test that after giving up on a downstairs, setting it to faulted, and
+    /// letting it reconnect, live repair occurs. Check that each extent is
+    /// repaired with the correct source, and that extent limits are honoured if
+    /// additional IO comes through.
+    #[tokio::test]
+    async fn test_successful_live_repair() {
+        let harness = Arc::new(TestHarness::new().await);
+
+        let (jh1, mut ds1_messages) =
+            harness.ds1().await.spawn_message_receiver().await;
+        let (_jh2, mut ds2_messages) =
+            harness.ds2.spawn_message_receiver().await;
+        let (_jh3, mut ds3_messages) =
+            harness.ds3.spawn_message_receiver().await;
+
+        // Send 1200 jobs. Flow control will kick in at 100 messages, so we need
+        // to be sending read responses while reads are being sent. After 1000
+        // jobs, the Upstairs will set ds1 to faulted, and send it no more work.
+        let mut job_ids = Vec::with_capacity(1200);
+
+        for i in 0..1200 {
+            info!(harness.log, "sending read {}", i);
+
+            {
+                let harness = harness.clone();
+
+                // We must tokio::spawn here because `read` will wait for the
+                // response to come back before returning
+                tokio::spawn(async move {
+                    let buffer = Buffer::new(512);
+                    harness
+                        .guest
+                        .read(Block::new_512(0), buffer)
+                        .await
+                        .unwrap();
+                });
+            }
+
+            if i < 100 {
+                // Before flow control kicks in, assert we're seeing the read
+                // requests
+                assert!(matches!(
+                    ds1_messages.recv().await.unwrap(),
+                    Message::ReadRequest { .. },
+                ));
+            } else {
+                // After flow control kicks in, we shouldn't see any more
+                // messages
+                assert!(matches!(
+                    ds1_messages.try_recv(),
+                    Err(TryRecvError::Empty)
+                ));
+            }
+
+            match ds2_messages.recv().await.unwrap() {
+                Message::ReadRequest { job_id, .. } => {
+                    // Record the job ids of the read requests
+                    job_ids.push(job_id);
+                }
+
+                _ => panic!("saw non read request!"),
```

Comment:
I thought I saw somewhere you set the flush_limit to be some large number
so the auto flush won't trip us up here, right?

Reply:
Ah, yes, I see it commented about below